### PR TITLE
Update in CHAP creds handling

### PR DESCRIPTION
**Pull request by sijeesh**
Store CHAP username and password in a secret and pass the secret name and namespace through storage class

### CODE REVIEW

**Review by datamattsson — COMMENTED** [review]
It passes CSI e2e tests, both with and without CHAP details.

However, if the CHAP credential is changed, the new credentail is leaked on the controller driver:

```
time="2024-04-23T00:50:20Z" level=error msg="Error controller publishing volume pvc-be32e35c-91b9-4495-9cc5-fde775da295d, err: rpc error: code = Internal desc = Failed to add ACL to volume pvc-be32e35c-91b9-4495-9cc5-fde775da295d for node &{ thirty-cpu1 e6d03e85-3685-179a-6837-6c62fcd100ed [0xc00091c2c0] [0xc00091c380 0xc00091c390 0xc00091c3a0 0xc00091c3b0] [] holler 1234567890abcdef} via CSP, err: Request failed with status code 404 and errors Error code (Not Found) and message (HTTP error response from backend {\"code\":17,\"desc\":\"host does not exist\"})" file="controller_server.go:729"
time="2024-04-23T00:50:20Z" level=error msg="GRPC error: rpc error: code = Internal desc = Failed to add ACL to volume pvc-be32e35c-91b9-4495-9cc5-fde775da295d for node &{ thirty-cpu1 e6d03e85-3685-179a-6837-6c62fcd100ed [0xc00091c2c0] [0xc00091c380 0xc00091c390 0xc00091c3a0 0xc00091c3b0] [] holler 1234567890abcdef} via CSP, err: Request failed with status code 404 and errors Error code (Not Found) and message (HTTP error response from backend {\"code\":17,\"desc\":\"host does not exist\"})" file="utils.go:73"
```

The username is "holler" and password "1234567890abcdef" in this example.

File: pkg/flavor/kubernetes/flavor.go [hunk 1]
```diff
@@ -932,3 +912,25 @@ func (flavor *Flavor) isVolumeAttachedToPod(pod *v1.Pod, va *storage_v1.VolumeAt
...
+	var chapSecret map[string]string
+	if volumeContext[chapSecretNameKey] != "" && volumeContext[chapSecretNamespaceKey] != "" {
+		chapSecretName := volumeContext[chapSecretNameKey]
+		chapSecretNamespace := volumeContext[chapSecretNamespaceKey]
+		log.Infof("Found CHAP secret %s secret namespace %s", chapSecretName, chapSecretNamespace)
+
+		var err error
+		chapSecret, err = flavor.GetCredentialsFromSecret(chapSecretName, chapSecretNamespace)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to read CHAP credentials from secret name %s secret namespace %s: %v", chapSecretName, chapSecretNamespace, err)
```
[the remark this answers is not in the document]
Comment (sijeesh) [reply pkg/flavor/kubernetes/flavor.go:929]: Tested it by removing secret object, here is the log from my env,

time="2024-04-23T07:39:31Z" level=trace msg="<<<<<< GetNodeInfo" file="flavor.go:338"
time="2024-04-23T07:39:31Z" level=trace msg=">>>>> GetChapCredentialsFromSecret" file="flavor.go:917"
time="2024-04-23T07:39:31Z" level=info msg="Found CHAP secret chap-secret secret namespace hpe-storage" file="flavor.go:924"
time="2024-04-23T07:39:31Z" level=trace msg=">>>>> GetCredentialsFromSecret, name: chap-secret, namespace: hpe-storage" file="flavor.go:616"
time="2024-04-23T07:39:31Z" level=trace msg="<<<<< GetCredentialsFromSecret" file="flavor.go:621"
time="2024-04-23T07:39:31Z" level=trace msg="<<<<< GetChapCredentialsFromSecret" file="flavor.go:929"
time="2024-04-23T07:39:31Z" level=error msg="Failed to check CHAP credentials availability in the volume context: failed to read CHAP credentials from secret name chap-secret secret namespace hpe-storage: error getting secret chap-secret in namespace hpe-storage: secrets \"chap-secret\" not found" file="controller_server.go:827"
time="2024-04-23T07:39:31Z" level=trace msg=">>>>> GetStorageProvider" file="driver.go:348"
time="2024-04-23T07:39:31Z" level=trace msg=">>>>> CreateCredentials" file="storage_provider.go:66"
time="2024-04-23T07:39:31Z" level=trace msg="<<<<< CreateCredentials" file="storage_provider.go:126"
time="2024-04-23T07:39:31Z" level=trace msg="Storage provider already exists. Returning it." file="driver.go:368"

**Comment by sijeesh** [comment]
> It passes CSI e2e tests, both with and without CHAP details.
> 
> However, if the CHAP credential is changed, the new credentail is leaked on the controller driver:
> 
> ```
> time="2024-04-23T00:50:20Z" level=error msg="Error controller publishing volume pvc-be32e35c-91b9-4495-9cc5-fde775da295d, err: rpc error: code = Internal desc = Failed to add ACL to volume pvc-be32e35c-91b9-4495-9cc5-fde775da295d for node &{ thirty-cpu1 e6d03e85-3685-179a-6837-6c62fcd100ed [0xc00091c2c0] [0xc00091c380 0xc00091c390 0xc00091c3a0 0xc00091c3b0] [] holler 1234567890abcdef} via CSP, err: Request failed with status code 404 and errors Error code (Not Found) and message (HTTP error response from backend {\"code\":17,\"desc\":\"host does not exist\"})" file="controller_server.go:729"
> time="2024-04-23T00:50:20Z" level=error msg="GRPC error: rpc error: code = Internal desc = Failed to add ACL to volume pvc-be32e35c-91b9-4495-9cc5-fde775da295d for node &{ thirty-cpu1 e6d03e85-3685-179a-6837-6c62fcd100ed [0xc00091c2c0] [0xc00091c380 0xc00091c390 0xc00091c3a0 0xc00091c3b0] [] holler 1234567890abcdef} via CSP, err: Request failed with status code 404 and errors Error code (Not Found) and message (HTTP error response from backend {\"code\":17,\"desc\":\"host does not exist\"})" file="utils.go:73"
> ```
> 
> The username is "holler" and password "1234567890abcdef" in this example.

Thanks for checking this, updated the logs to avoid logging full data of the node.

File: pkg/flavor/kubernetes/flavor.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (datamattsson) [reply pkg/flavor/kubernetes/flavor.go:929]: Ah, my mistake, the logic is wrong elsewhere. The symptom I'm seeing is that if the Secret can't be found, we fail silently and the proceed with the attachment without CHAP.

[the remark this answers is not in the document]
Comment (sijeesh) [reply pkg/flavor/kubernetes/flavor.go:929]: can you please share the steps you tried, would like to reproduce in my env and check it.

**Review by dileepds — APPROVED** [review]
LGTM

**Review by datamattsson — APPROVED** [review]
- Passes e2e tests with and without CHAP secret on 3PAR
- Passes e2e tests with CHAP secret on Nimble
- Fails if CHAP secret is missing
- Fails if CHAP password is invalid with misleading errors on the Pods failing.
   - Error seen: `  Warning  FailedAttachVolume  32s (x9 over 2m44s)  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-1bcdcdb5-a137-4f0e-9fd2-d7cfef219d40" : rpc error: code = Internal desc = Failed to add ACL to volume 0655a648f954b5cddb000000000000000000004a1e for node  via CSP, err: Request failed with status code 400 and errors`
   - Actual relevant error (CSP log): `00:31:25 ERROR [co.ni.hi.cs.se.er.GroupMgmtServerExceptionMapper]] (executor-thread-114) Bad request: com.nimblestorage.nos.groupmgmt.GroupMgmtServerException: HTTP 400 OK ([{ code: SM_http_bad_request,
 text: The request could not be understood by the server.,
 arguments: null}
, { code: SM_invalid_arg_value,
 text: Invalid value '****' specified for the argument 'password'.,
 arguments: {arg=password, value=****}}
])` 
  We'll just have to document our way out of it.
- Does NOT fail if keys are messed up in the Secret (I see this as optional)

**Pull request merged**